glommio/src/executor/stall.rs

```diff
@@ -0,0 +1,543 @@
+        &self,
+        _queue_handle: TaskQueueHandle,
+        max_expected_runtime: Duration,
+    ) -> Option<Duration> {
+        Some(
+            Duration::from_millis(
+                (max_expected_runtime.as_millis() as f64 * self.overage_ratio) as u64,
+            )
+            .max(self.min_duration),
+        )
+    }
+
+    /// The default signal is [`nix::libc::SIGUSR1`].
+    fn signal(&self) -> u8 {
+        self.signal
+    }
+
+    /// The default stall reporting mechanism is to log a warning.
+    fn stall(&self, detection: StallDetection<'_>) {
+        log::warn!("{}", detection);
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct StallDetector {
+    timer: Arc<sys::timerfd::TimerFd>,
+    stall_handler: Box<dyn StallDetectionHandler + 'static>,
+    timer_handler: Option<JoinHandle<()>>,
+    id: usize,
+    terminated: Arc<AtomicBool>,
+    signal_id: signal_hook::SigId,
+    // NOTE: we don't use signal_hook::low_level::channel as backtraces
+    // have too many elements
+    rx: crossbeam::channel::Receiver<backtrace::BacktraceFrame>,
+}
+
+impl StallDetector {
+    pub(crate) fn new(
+        executor_id: usize,
+        stall_handler: Box<dyn StallDetectionHandler + 'static>,
+    ) -> std::io::Result<StallDetector> {
+        let timer = Arc::new(
+            sys::timerfd::TimerFd::new(
+                sys::timerfd::ClockId::CLOCK_MONOTONIC,
+                sys::timerfd::TimerFlags::empty(),
+            )
+            .map_err(std::io::Error::from)?,
+        );
+        let terminated = Arc::new(AtomicBool::new(false));
+        let sig = stall_handler.signal();
+        let (tx, rx) = crossbeam::channel::bounded(1 << 10);
+        let timer_handler =
+            StallDetector::install_trigger(terminated.clone(), timer.clone(), sig.into());
+        let signal_id = StallDetector::install_handler(tx, sig)?;
+
+        Ok(Self {
+            timer,
+            timer_handler: Some(timer_handler),
+            stall_handler,
+            id: executor_id,
+            terminated,
+            signal_id,
+            rx,
+        })
+    }
+
+    fn install_handler(
+        tx: crossbeam::channel::Sender<backtrace::BacktraceFrame>,
+        signal: u8,
+    ) -> std::io::Result<signal_hook::SigId> {
+        let exec_thread = thread::current().id();
+        unsafe {
+            let signal_id = signal_hook::low_level::register(signal.into(), move || {
+                // Bail if we can't send or if we've gotten a signal
+                // from an unexpected thread (i.e., a signal targeting the process)
+                if tx.is_full() || thread::current().id() != exec_thread {
+                    return;
+                }
+
+                backtrace::trace_unsynchronized(|frame| {
+                    tx.try_send(backtrace::BacktraceFrame::from(frame.clone()))
+                        .is_ok()
+                });
+            })?;
+            //.map_err::<io::Error, GlommioError<()>>(GlommioError::from)?;
```

Comment:
nit

Reply:
Btw I think you can just return the directly rather than using `?` + `Ok(signal_id)`